ambari-server/src/main/java/org/apache/ambari/server/state/cluster/ClustersImpl.java

```diff
@@ -169,41 +169,143 @@ public ClustersImpl(ClusterDAO clusterDAO, ClusterFactory clusterFactory, HostDA
+    if (clustersById == null) {
+      safelyLoadClustersAndHosts();
+    }
+    return clustersById;
+  }
+
+  /**
+   * Gets the internal hosts-by-name map, ensuring that the relevant data has been previously initialized.
+   *
+   * @return a map of the requested data
+   */
+  private ConcurrentHashMap<String, Host> getHostsByName() {
+    if (hostsByName == null) {
+      safelyLoadClustersAndHosts();
+    }
+    return hostsByName;
+  }
+
+  /**
+   * Gets the internal hosts-by-id map, ensuring that the relevant data has been previously initialized.
+   *
+   * @return a map of the requested data
+   */
+  private ConcurrentHashMap<Long, Host> getHostsById() {
+    if (hostsById == null) {
+      safelyLoadClustersAndHosts();
+    }
+    return hostsById;
+  }
+
+  /**
+   * Gets the internal host/clusters map, ensuring that the relevant data has been previously initialized.
+   *
+   * @return a map of the requested data
+   */
+  private ConcurrentHashMap<String, Set<Cluster>> getHostClustersMap() {
+    if (hostClustersMap == null) {
+      safelyLoadClustersAndHosts();
+    }
+    return hostClustersMap;
+  }
+
+  /**
+   * Gets the internal cluster/hosts map, ensuring that the relevant data has been previously initialized.
+   *
+   * @return a map of the requested data
+   */
+  private ConcurrentHashMap<String, Set<Host>> getClusterHostsMap() {
+    if (clusterHostsMap1 == null) {
+      safelyLoadClustersAndHosts();
+    }
+    return clusterHostsMap1;
+  }
+
+  /**
+   * Safely initializes all of the in-memory state collections that this class.
+   * <p>
+   * This method is synchronized so that the data can be loaded only if needed.  If most than one of
+   * the relevant get methods are concurrently invoked, this method should ensure that {@link #loadClustersAndHosts()}
+   * will be entered once.  Subsequent calls will bypass {@link #loadClustersAndHosts()} since the
+   * relevant variables will no longer be null.
+   *
+   * @see #getClustersByName()
+   * @see #getClustersById()
+   * @see #getHostsByName()
+   * @see #getHostsById()
+   * @see #getClusterHostsMap()
+   * @see #getHostClustersMap()
+   */
+  synchronized private void safelyLoadClustersAndHosts() {
+    if (clustersByName == null || clustersById == null ||
+        hostsByName == null || hostsById == null ||
+        hostClustersMap == null || clusterHostsMap1 == null) {
+      loadClustersAndHosts();
+    }
+  }
+
+  /**
+   * Initializes all of the in-memory state collections that this class
+   * unfortunately uses.
+   * <p>
+   * This method should be called only once, when the data is first needed.
    * <p/>
    * Because some of these stateful initializations may actually reference this
    * {@link Clusters} instance, we must do this after the object has been
    * instantiated and injected.
    */
-  @Inject
   @Transactional
-  void loadClustersAndHosts() {
+  private void loadClustersAndHosts() {
```

Comment:
Guice AOP interceptors don't work on private methods.
This means that Transactional annotation will be ignored.
Please check what is correct behavior here.

Reply:
Is a transaction even relevant here? The method does a bunch of DAO reads.  @jonathan-hurley?

Reply:
Theres RequiresSession annotation to mimic read-only transaction behavior.
Basically if session is not already opened on method call, each DAO read will create own session.
Not really relevant here since this method is not called frequently.

Reply:
Oh.. true. I can add the RequiresSession annotation if we think it is needed.  I removed `@Transactional`

Reply:
That `@Transactional` was added a long, long time ago by Myroslav - I agree it's not needed.